feat: enhance BLS testing framework and Ethereum validation utilities

yilmazbahadir opened the pull request:
## Summary

Enhances GridPlus SDK testing infrastructure with improved BLS signature testing and Ethereum validation utilities.

## ✨ Key Changes

### 🔐 BLS Testing Improvements
- Better firmware compatibility detection for BLS12-381 support
- Graceful test skipping with clear diagnostics
- Enhanced error handling and debugging information

### 🔍 Ethereum Validation
- New validation functions for addresses and transactions
- Comprehensive unit test coverage
- Robust error handling with TypeScript support

### 🧪 Testing Infrastructure
- Enhanced determinism testing with improved test vectors
- Better TypeScript configuration (`tsconfig.build.json`)
- Improved vitest setup and type definitions

## 🧪 Testing

- ✅ All signing tests are green, run `e2e-sign` with
`CI=1 DEBUG_SIGNING=1 baseUrl=http://127.0.0.1:3000 DEVICE_ID=SD0001 PASSWORD=12345678 PAIRING_SECRET=12345678 ENC_PW=12345678 APP_NAME=lattice-manager pnpm e2e-sign`
<img width="526" height="93" alt="image" src="https://github.com/user-attachments/assets/d49af79b-7846-4c43-874a-a55290b53f39" />

- ✅ New unit tests for Ethereum validation
- ✅ Enhanced BLS testing with better error handling
- ✅ TypeScript compilation successful

yilmazbahadir committed:
feat: enhance bls testing and ethereum validation

- Add comprehensive BLS test improvements with better firmware compatibility checks
- Implement Ethereum validation utilities with robust error handling
- Add unit tests for Ethereum validation functions
- Enhance determinism testing with improved test vectors
- Add TypeScript build configuration for better type checking
- Update package.json with new test dependencies
- Improve test environment setup with vitest configuration

Key improvements:
- Better BLS firmware version detection and feature checking
- Enhanced Ethereum address and transaction validation
- More comprehensive test coverage for edge cases
- Improved TypeScript configuration for development

yilmazbahadir committed:
fix: remove package manager version

yilmazbahadir committed:
fix: lint

netbonus commented on src/__test__/unit/ethereum.validate.test.ts:
```diff
@@ -0,0 +1,92 @@
+import { buildFirmwareConstants, DEFAULT_SIGNER } from '../utils/builders';
+import { HARDENED_OFFSET } from '../../constants';
+
+const TEST_MNEMONIC =
```
do we need to use this mnemonic? would be nice to use a global one for all testing.

yilmazbahadir replied:
good idea! will do.

netbonus commented on src/ethereum.ts:
```diff
@@ -86,13 +86,37 @@ const validateEthereumMsgResponse = function (res, req) {
+      payloadForHashing,
       SignTypedDataVersion.V4,
     );
+    if (process.env.DEBUG_SIGNING) {
```
why not use `console.debug` ?

we could actually drop `console.debug` from the shipped bundle using tsup, if we want.

yilmazbahadir replied:
missed this one, I don't think we need this one.

netbonus commented on src/ethereum.ts:
```diff
@@ -86,13 +86,37 @@ const validateEthereumMsgResponse = function (res, req) {
     const digest = prehash ? prehash : encoded;
-    const chainId = parseInt(input.payload.domain.chainId, 16);
+    // Parse chainId - it could be a number, hex string, or decimal string
+    let chainId =
```
would it ever be a bigint?

yilmazbahadir replied:
yes, it might be. added below 🙌

netbonus commented on src/ethereum.ts:
```diff
@@ -103,6 +127,22 @@ const validateEthereumMsgResponse = function (res, req) {
 function convertBigNumbers(obj) {
```
gh is saying this isnt used?

yilmazbahadir replied:
removed.

netbonus commented on src/ethereum.ts:
```diff
@@ -116,6 +156,72 @@ function convertBigNumbers(obj) {
   }
 }
 
+function normalizeTypedDataForHashing(value: any): any {
```
is there a viem/ox function that does this?

yilmazbahadir replied:
> Based on my research, no, we should keep normalizeTypedDataForHashing rather than replacing it with viem/ox functions. Here's why:
> 
> 
>    Analysis:
> 
>   1. Viem's hashTypedData doesn't solve the problem
> 
>   - Viem's hashTypedData can hash EIP-712 data, but the documentation doesn't detail how it handles type normalization for bigint, BN instances,
>   buffers, etc.
>   - Even if we switched to viem's hashing, we'd still need normalization before passing data to it
> 
>   2. The SDK has a unique challenge
> 
>   The GridPlus SDK receives EIP-712 data from many different sources:
>   - dApps using ethers.js (which uses ethers.BigNumber)
>   - dApps using viem (which uses native bigint)
>   - dApps using web3.js (which might use BN from bn.js)
>   - Older dApps using bignumber.js
> 
>   Each library represents numbers differently, and normalizeTypedDataForHashing handles all of them.
> 
>   3. Dual-purpose encoding (lines 761-775, 800-803)
> 
>   The code creates TWO different payloads for the same EIP-712 data:
>   - Firmware payload: CBOR-encoded, requires BN (bignumber.js) instances
>   - Validation payload: For TypedDataUtils.eip712Hash, requires numbers/strings
> 
>   normalizeTypedDataForHashing specifically prepares data for the validation step.
> 
>   4. MetaMask's library is battle-tested
> 
>   Using @metamask/eth-sig-util for EIP-712 hashing is the de facto standard for wallet integrations. Switching would be risky.
> 
>   Recommendation:
> 
>   Keep normalizeTypedDataForHashing as-is. It serves a legitimate, specific purpose that viem/ox don't directly address. The PR review comment was
>    correct to question whether viem has this functionality, but after analysis, the custom normalization is justified for this SDK's use case.

netbonus commented on src/ethereum.ts:
```diff
@@ -116,6 +156,72 @@ function convertBigNumbers(obj) {
+  return value;
+}
+
 const buildEthereumTxRequest = function (data) {
```
at some point we want to push this logic into viem

netbonus commented on src/ethereum.ts:
```diff
@@ -765,25 +885,41 @@ function buildEIP712Request(req, input) {
-  input.payload.message = parseEIP712Msg(
+  // IMPORTANT: Create a new object for the validation payload instead of modifying input.payload
+  // in place, so that validation uses the correctly formatted data
+  const validationPayload = JSON.parse(JSON.stringify(data));
```
that seems like a fun bug to find 😅

yilmazbahadir committed:
refactor(ethereum): centralize test constants and improve EIP-712 type handling

- Centralize test mnemonics in testConstants.ts to reduce duplication
- Remove unused DEBUG_SIGNING environment variable logging
- Add bigint type support for chainId in EIP-712 validation
- Remove dead code: unused convertBigNumbers function

Addresses all review feedback on PR #620:
- Eliminates TEST_MNEMONIC duplication across 3 test files
- Removes debug logging in favor of cleaner code
- Ensures compatibility with modern libraries (viem, ethers v6) using bigint
- Cleans up 32 lines of unused convertBigNumbers utility function

yilmazbahadir committed:
fix: use correct hash type in generic signing and improve multipart signing

Fixes generic signing to respect the request's hashType (SHA256 vs keccak256) when recovering signatures, preventing incorrect signature validation. Also improves multipart signing by adding better error handling for missing extra data payloads and fixing the hasExtraPayloads flag calculation.

yilmazbahadir committed:
fix: lint

yilmazbahadir commented on src/functions/sign.ts:
```diff
@@ -114,23 +116,31 @@ export const encodeSignRequest = ({
+    reqPayload = Buffer.concat([nextCode, nextExtraPayload]);
     schema = LatticeSignSchema.extraData;
+    hasExtraPayloads = Number(
+      (typedCachedData.extraDataPayloads?.length ?? 0) > 0,
```
it was always [checking](https://github.com/GridPlus/gridplus-sdk/pull/620/files#diff-ad159ab6c96ea0caf4a3556f3551c429f898423c85003d9e9f6fa1cf22a33f20L130)  `typedRequestData.extraDataPayloads`, but when we're in the second call (with cachedData &&  nextCode), we should be checking `typedCachedData.extraDataPayloads` AFTER the shift.
